### PR TITLE
feat(curriculum): add chapters and modules

### PR DESCRIPTION
**Pull request by ojeytonwilliams**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Okay, this does a couple of things: it adds the `chapter` and `module` properties to the challenge object and it lets us control the order of the blocks via one config file. 

The `chapter` and `module` properties are included so that the client can access them when making GraphQL queries for challenges. This should be enough information to let us build the new UI. Though, we can always add more if necessary.

The rationale for controlling the order via a config file is partly because the information is already contained in the file (blocks had to go in and an array was the natural data structure), partly because it makes it trivial to change that order and partly because it gives us a nice high-level overview of what goes into a superblock.

The downside is that it adds a little extra complexity, because we have to support the existing approach (order property in the meta.json) for the time being. It's not very complicated, though, so even if we have to support both for years it should be okay.

Also, I ended up putting some blocks in the wrong modules, because it wasn't clear where to put them. The asynchronous-javascript, functional-programming and react modules all have some extra, since they seemed like the best fit.

Finally, I moved the typescript block to the correct place.

- Closes https://github.com/freeCodeCamp/fCC10/issues/21.
- Closes https://github.com/freeCodeCamp/fCC10/issues/36.

### TODO
- [x] ~rename superblock once https://github.com/freeCodeCamp/freeCodeCamp/pull/56694/ is in~ on second thought, it doesn't matter which order these go in.
- [x] derive block order from structure config, not `order` property
- [x] update challenge schema (with apologies to the mobile team)
- [x] add a structure-schema analogous to the meta-schema
- [x] update Gatsby schemaCustomization to include `chapter` and `module`

### CODE REVIEW

**Review by huyenltnguyen — COMMENTED** [review]
LGTM.

Quick question: We will have a capstone project and an exam that won't belong to any chapters, so I'm wondering how `full-stack.json` would look. I'm assuming it'd be:

```json
{
  "chapters":  [ // ... ],
  "capstone-project": {
    "dashedName": "fsd-capstone-project" 
  },
  "exam": {
    "dashedName": "fsd-exam" 
  }
}
```
?

File: curriculum/challenges/_meta/lab-telephone-number-validator/meta.json [hunk 1]
```diff
@@ -5,9 +5,7 @@
     "isUpcomingChange": true,
     "usesMultifileEditor": true,
     "dashedName": "lab-telephone-number-validator",
-    "order": 246,
-    "superBlock": "front-end-development",
+  "superBlock": "front-end-development",
```
[the remark this answers is not in the document]
Comment (moT01) [reply curriculum/challenges/_meta/lab-telephone-number-validator/meta.json:8]: :check:

File: curriculum/challenges/_meta/lab-url-shortener-microservice/meta.json [hunk 2]
```diff
@@ -4,8 +4,7 @@
     "blockLayout": "link",
     "isUpcomingChange": true,
     "dashedName": "lab-url-shortener-microservice",
-    "order": 374,
-    "superBlock": "front-end-development",
+  "superBlock": "front-end-development",
```
[the remark this answers is not in the document]
Comment (moT01) [reply curriculum/challenges/_meta/lab-url-shortener-microservice/meta.json:7]: ✅

**Comment by moT01** [comment]
> We will have a capstone project and an exam that won't belong to any chapters

I'm not sure exactly. I'm thinking that within the JSON, they will have a chapter...

```
{
  "chapters":  [
    ...,
    {
      "dashedName": "final-exam",
      "modules": [
        {
          "dashedName": "final-exam",
          "blocks": [{ "dashedName": "final-exam" }]
        }
      ]
    }
  ]
}
```

That probably isn't the best. We would then need some client stuff to display that chapter as its own challenge or something. Or perhaps we can modify the structure - maybe give it a chapterType.

File: curriculum/superblock-structure/full-stack.json [hunk 3]
```diff
@@ -0,0 +1,486 @@
+{
+  "chapters": [
+    {
+      "dashedName": "html",
+      "modules": [
+        {
+          "dashedName": "getting-started-with-freecodecamp",
+          "blocks": [{ "dashedName": "lecture-what-is-html" }]
```
[the remark this answers is not in the document]
Comment (moT01) [reply curriculum/superblock-structure/full-stack.json]: ✅

**Comment by moT01** [comment]
I added the blocked label here because I want to rebase and fix up any curriculum stuff that was added right before this gets merged - to avoid tests failing on main. It could still use some reviews and approvals.

**Review by huyenltnguyen — APPROVED** [review]
LGTM 👍 Thanks a lot for the work!

**Review by moT01 — APPROVED** [review]

**Comment by moT01** [comment]
Open curriculum PR's that add a new block will need to remove the order from the `meta.json` file and add the block where it fits in the `full-stack.json` file

**Pull request merged**